mailpoet/lib/Automation/Engine/Control/StepHandler.php

```diff
@@ -78,55 +78,70 @@ public function initialize(): void {
 
   /** @param mixed $args */
   public function handle($args): void {
-    // TODO: args validation
-    if (!is_array($args)) {
+    // TODO: better args validation
+    if (!is_array($args) || !isset($args['automation_run_id']) || !array_key_exists('step_id', $args)) {
       throw new InvalidStateException();
     }
 
-    // Action Scheduler catches only Exception instances, not other errors.
-    // We need to convert them to exceptions to be processed and logged.
+    $runId = (int)$args['automation_run_id'];
+    $stepId = (string)$args['step_id'];
+    $runNumber = (int)($args['run_number'] ?? 1);
+
+    // BC — complete automation run if "step_id" is empty (was nullable in the past)
```

Comment:
For being backwards compatible. Do you have an idea on how long we want to maintain this?

Reply:
Most users that already use automations will probably update the plugin sooner than later, and then very quickly all of the steps with `null` should disappear. Usually, those were scheduled to "immediately", I think. My guess is that a couple of months should be fine, but it's not a huge piece of code, so I guess it can stay for a while.